controllers/templatesync/template_sync.go

```diff
@@ -523,11 +535,97 @@ func (r *PolicyReconciler) Reconcile(ctx context.Context, request reconcile.Requ
 		policySystemErrorsCounter.WithLabelValues(instance.Name, "", "client-error").Inc()
 	}
 
+	err = r.cleanUpExcessTemplates(ctx, dClient, *instance, templateNames)
+	if err != nil {
+		resultError = err
+		reqLogger.Error(resultError, "Error cleaning up templates")
+	}
+
 	reqLogger.Info("Completed the reconciliation")
 
 	return reconcile.Result{}, resultError
 }
 
+// cleanUpExcessTemplates compares existing policy templates on the cluster to those contained in the policy,
+// and deletes those that have been renamed or removed from the parent policy
+func (r *PolicyReconciler) cleanUpExcessTemplates(
+	ctx context.Context,
+	dClient dynamic.Interface,
+	instance policiesv1.Policy,
+	templateNames []string,
+) error {
+	crdLabelSelector := labels.SelectorFromSet(map[string]string{PolicyTypeLabel: "template"})
+	crdsv1 := extensionsv1.CustomResourceDefinitionList{}
+	tmplGVRs := []schema.GroupVersionResource{}
+
+	// build list of GVRs for templates to check the parent label on
+	err := r.List(ctx, &crdsv1, &client.ListOptions{LabelSelector: crdLabelSelector})
+	if err == nil {
+		for _, crd := range crdsv1.Items {
+			if len(crd.Spec.Versions) > 0 {
+				tmplGVRs = append(tmplGVRs, schema.GroupVersionResource{
+					Group:    crd.Spec.Group,
+					Resource: crd.Spec.Names.Plural,
+					Version:  crd.Spec.Versions[0].Name,
+				})
+			}
+		}
+	} else if err.Error() == "no matches for kind \"CustomResourceDefinition\" in version \"apiextensions.k8s.io/v1\"" {
```

Comment:
Do you know if there is anything from `k8s.io/apimachinery/pkg/api/errors` that we'd be able to use instead of string matching? I'm just worried that this is a bit brittle.

Reply:
I looked on friday and wasn't able to find a function in `k8s.io/apimachinery/pkg/api/errors` for this scenario. If you're worried about it I can change that block back to an `else` and have the error returned not be version-specific.

Reply:
@JustinKuli what do you think?

Reply:
I found https://pkg.go.dev/k8s.io/apimachinery@v0.26.1/pkg/api/meta#IsNoMatchError, it's slightly more general than what this checks for (which could be good). It's hard to tell if the string matching could break in the future.

Reply:
nice find @JustinKuli ! I didn't think to check the meta library. I'll add that instead of the string match